README.md

[![Arduino CI](https://github.com/RobTillaart/Gauss/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/Gauss/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/Gauss/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/Gauss/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/Gauss/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/Gauss/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/Gauss.svg?maxAge=3600)](https://github.com/RobTillaart/Gauss/releases)


# Gauss

Library for the Gauss probability math. (normal distribution).


## Description

Gauss is an experimental Arduino library to approximate the probability that a value is 
smaller or larger than a given value.
These under the premises of a Gaussian distribution with parameters **mean** and **stddev** 
(a.k.a. average / mu / µ and standard deviation / sigma / σ).
If these parameters are not given, 0 and 1 are used by default (normalized Gaussian distribution).

The values are approximated with **MultiMap()** using a 34 points interpolated lookup.
Therefore the **MultiMap** library need to be downloaded too (see related below).
The number of lookup points might chance in the future.

Return values are given as a float 0.0 .. 1.0, if one needs percentages, multiply by 100.0.


#### Accuracy

The lookup table has 34 points with 8 decimals.  
This matches the precision of float data type.
Do not expect a very high accuracy / precision as interpolation is linear.
For many applications this accuracy is sufficient.

Values of the table are calculated with ```NORM.DIST(mean, stddev, x, true)```.

Note: 0.1.0 was 32 points 4 decimals. Need to investigate reduction of points.


#### Applications

- use as a filter? do not allow > 3 sigma
- compare historic data to current data
- compare population data with individual


#### Character

|  parameter  |  name  |  ALT-code  |  char |
|:-----------:|:------:|:----------:|:-----:|
|  mean       |  mu    |  ALT-230   |   µ   |
|  stddev     | sigma  |  ALT-229   |   σ   |

- https://altcodesguru.com/greek-alt-codes.html


#### Related

- https://en.wikipedia.org/wiki/Normal_distribution
- https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_probability/bs704_probability9.html
- https://github.com/RobTillaart/Multimap
- https://github.com/RobTillaart/Statistic  (more stat links there).


## Interface

```cpp
#include Gauss.h
```


#### Base

- **Gauss()** constructor. Uses mean = 0 and stddev = 1 by default.
- **bool begin(float mean = 0, float stddev = 1)** set the mean and stddev.
Returns true if stddev > 0 which should be so.
Returns false if stddev <= 0, which could be a user choice.
Note that if ```stddev == 0```, probabilities cannot be calculated 
as the distribution is not Gaussian.
The default values (0,1) gives the normalized Gaussian distribution.
**begin()** can be called at any time to change the mean or stddev.
- **float getMean()** returns current mean.
- **float getStddev()** returns current stddev.


#### Probability

Probability functions return NAN if stddev == 0.  
Multiply probabilities by 100.0 to get the value as a percentage.

- **float P_smaller(float f)** returns probability **P(x < f)**.
A.k.a. **CDF()** Cumulative Distribution Function.
- **float P_larger(float f)** returns probability **P(x > f)**.
As the distribution is continuous **P_larger(f) == 1 - P_smaller(f)**.
- **float P_between(float f, float g)** returns probability **P(f < x < g)**.
- **float P_equal(float f)** returns probability **P(x == f)**.
This uses the bell curve formula.
- **float P_outside(float f, float g)** returns probability **P(x < f) + P(g < x)**.
  - **P_outside() = 1 - P_between()**


#### Normalize

- **float normalize(float f)** normalize a value to normalized distribution.
E.g if mean == 50 and stddev == 14, then 71 ==> +1.5 sigma.
Is equal to number of **stddevs()**.
- **float denormalize(float f)** reverses normalize(). 
What value would have a deviation of 1.73 stddev.
- **float stddevs(float f)** returns the number of stddevs from the mean.
Identical to **normalize()**.


#### Other

- **float bellCurve(float f)** returns probability **P(x == f)**.


## Performance

Indicative numbers for 1000 calls, timing in micros.

Arduino UNO, 16 MHz, IDE 1.8.19

|  function     |  0.1.0   |  0.1.1   |  0.2.0   |  notes  |
|:--------------|:--------:|:--------:|:--------:|:--------|
|  P_smaller    |  375396  |  365964  |  159536  |
|  P_larger     |  384368  |  375032  |  169056  |
|  P_between    |  265624  |  269176  |  150148  |
|  normalize    |   44172  |   23024  |   23024  |
|  bellCurve    |  255728  |  205460  |  192524  |
|  approx.bell  |  764028  |  719184  |  333172  |  see examples 


ESP32, 240 MHz, IDE 1.8.19

|  function     |  0.1.0   |  0.1.1   |  0.2.0   |  notes  |
|:--------------|:--------:|:--------:|:--------:|:--------|
|  P_smaller    |    -     |    4046  |    1498  |
|  P_larger     |    -     |    4043  |    1516  |
|  P_between    |    -     |    3023  |    1569  |
|  normalize    |    -     |     592  |     585  |
|  bellCurve    |    -     |   13522  |   13133  |
|  approx.bell  |    -     |    7300  |    2494  | 


## Future

#### Must

- documentation


#### Should


#### Could

- add examples
- add unit tests
- **VAL(probability = 0.75)** ==>  134 whatever
  - Returns the value of the distribution for which the **CDF()** is at least probability.
  - Inverse of **P_smaller()** (how? binary search)
  

#### Won't (unless requested)

- equality test Gauss objects
- does the stddev needs to be positive? Yes.
  - what happens if negative values are allowed? P curve is reversed.
- move code to .cpp file? (rather small lib).
- **void setMean(float f)** can be done with begin()
- **void setStddev(float f)** can be done with begin()
- optimize accuracy
  - (-6 .. 0) might be more accurate (significant digits)?


CHANGELOG.md
# Change Log Gauss

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2023-07-07
- improve performance => faster lookup.
  - removed __z[] saves 136 bytes RAM
- remove MultiMap dependency.
- - remove **\_stddev** as **\_reciprokeSD** holds same information.
- add **float P_outside(float f, float g)** 
  - returns **P(x < f) + P(g < x)** under condition (f < g).
- add **float denormalize(float value)** (reverse normalize).
- add examples DS18B20 and HX711
- minor edits.


----

## [0.1.1] - 2023-07-07
- improve performance => reciprokeSD = 1.0/stddev
- update readme.md
  - add performance section (UNO / ESP32)
  - elaborated future section
- generated a more precise lookup table (8 decimals)
- update unit tests
- add default parameters to **bool begin(float mean = 0, float stddev = 1)**
  - allow negative **stddev** but return false if stddev <= 0.
- add **float getMean()** convenience function.
- add **float getStdDev()** convenience function.
- clean up a bit

## [0.1.0] - 2023-07-06
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-07
// PURPOSE: unit tests for the Gauss library
//          https://github.com/RobTillaart/Gauss
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)


#include <ArduinoUnitTests.h>

#include "Arduino.h"
#include "Gauss.h"


unittest_setup()
{
    fprintf(stderr, "GAUSS_LIB_VERSION: %s\n", (char *) GAUSS_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constructor)
{
  Gauss G;
 
  assertEqualFloat(0.0, G.getMean(), 0.0001);
  assertEqualFloat(1.0, G.getStdDev(), 0.0001);
  assertEqualFloat(0.5, G.P_smaller(0), 0.0001);

  G.begin(10, 3);
  assertEqualFloat(10.0, G.getMean(), 0.0001);
  assertEqualFloat(3.0,  G.getStdDev(), 0.0001);
}


unittest(test_P_smaller)
{
  Gauss G;

  G.begin(0, 1);

  assertEqualFloat(0.0000, G.P_smaller(-6.0), 0.0001);
  assertEqualFloat(0.0001, G.P_smaller(-4.0), 0.0001);
  assertEqualFloat(0.0013, G.P_smaller(-3.0), 0.0001);
  assertEqualFloat(0.0228, G.P_smaller(-2.0), 0.0001);
  assertEqualFloat(0.1587, G.P_smaller(-1.0), 0.0001);
  assertEqualFloat(0.5000, G.P_smaller(0.0), 0.0001);
  assertEqualFloat(0.8413, G.P_smaller(1.0), 0.0001);
  assertEqualFloat(0.9772, G.P_smaller(2.0), 0.0001);
  assertEqualFloat(0.9987, G.P_smaller(3.0), 0.0001);
  assertEqualFloat(0.9999, G.P_smaller(4.0), 0.0001);
  assertEqualFloat(1.0000, G.P_smaller(6.0), 0.0001);
}


unittest(test_P_larger)
{
  Gauss G;

  G.begin(0, 1);

  assertEqualFloat(0.9987, G.P_larger(-3.0), 0.0001);
  assertEqualFloat(0.9772, G.P_larger(-2.0), 0.0001);
  assertEqualFloat(0.8413, G.P_larger(-1.0), 0.0001);
  assertEqualFloat(0.5000, G.P_larger(0.0), 0.0001);
  assertEqualFloat(0.1587, G.P_larger(1.0), 0.0001);
  assertEqualFloat(0.0228, G.P_larger(2.0), 0.0001);
  assertEqualFloat(0.0013, G.P_larger(3.0), 0.0001);
}


unittest(test_P_between)
{
  Gauss G;

  G.begin(0, 1);

  assertEqualFloat(0.4987, G.P_between(-3.0, 0.0), 0.0001);
  assertEqualFloat(0.4772, G.P_between(-2.0, 0.0), 0.0001);
  assertEqualFloat(0.3413, G.P_between(-1.0, 0.0), 0.0001);
  assertEqualFloat(0.0000, G.P_between(0.0, 0.0), 0.0001);
  assertEqualFloat(0.3413, G.P_between(0.0, 1.0), 0.0001);
  assertEqualFloat(0.4772, G.P_between(0.0, 2.0), 0.0001);
  assertEqualFloat(0.4987, G.P_between(0.0, 3.0), 0.0001);
}


unittest(test_P_outside)
{
  Gauss G;

  G.begin(0, 1);

  assertEqualFloat(G.P_outside(-3.0, 0.0), 1 - G.P_between(-3.0, 0.0), 0.0001);
  assertEqualFloat(G.P_outside(-2.0, 0.0), 1 - G.P_between(-2.0, 0.0), 0.0001);
  assertEqualFloat(G.P_outside(-1.0, 0.0), 1 - G.P_between(-1.0, 0.0), 0.0001);
  assertEqualFloat(G.P_outside(0.0, 0.0),  1 - G.P_between(0.0, 0.0),  0.0001);
  assertEqualFloat(G.P_outside(0.0, 1.0),  1 - G.P_between(0.0, 1.0),  0.0001);
  assertEqualFloat(G.P_outside(0.0, 2.0),  1 - G.P_between(0.0, 2.0),  0.0001);
  assertEqualFloat(G.P_outside(0.0, 3.0),  1 - G.P_between(0.0, 3.0),  0.0001);
}


unittest(test_P_equal)
{
  Gauss G;

  G.begin(0, 1);

  assertEqualFloat(0.004432, G.P_equal(-3.0), 0.0001);
  assertEqualFloat(0.053991, G.P_equal(-2.0), 0.0001);
  assertEqualFloat(0.241971, G.P_equal(-1.0), 0.0001);
  assertEqualFloat(0.398942, G.P_equal(0.0), 0.0001);
  assertEqualFloat(0.241971, G.P_equal(1.0), 0.0001);
  assertEqualFloat(0.053991, G.P_equal(2.0), 0.0001);
  assertEqualFloat(0.004432, G.P_equal(3.0), 0.0001);
}


unittest(test_normailze)
{
  Gauss G;

  G.begin(100, 25);

  assertEqualFloat(-3.0, G.normalize(25), 0.0001);
  assertEqualFloat(-2.0, G.normalize(50), 0.0001);
  assertEqualFloat(-1.0, G.normalize(75), 0.0001);
  assertEqualFloat(0.0, G.normalize(100), 0.0001);
  assertEqualFloat(1.0, G.normalize(125), 0.0001);
  assertEqualFloat(2.0, G.normalize(150), 0.0001);
  assertEqualFloat(3.0, G.normalize(175), 0.0001);

  assertEqualFloat(25, G.denormalize(-3.0), 0.0001);
  assertEqualFloat(50, G.denormalize(-2.0), 0.0001);
  assertEqualFloat(75, G.denormalize(-1.0), 0.0001);
  assertEqualFloat(100, G.denormalize(0.0), 0.0001);
  assertEqualFloat(125, G.denormalize(1.0), 0.0001);
  assertEqualFloat(150, G.denormalize(2.0), 0.0001);
  assertEqualFloat(175, G.denormalize(3.0), 0.0001);
}


unittest_main()


//  -- END OF FILE --



Gauss.h
#pragma once
//
//    FILE: Gauss.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.0
// PURPOSE: Library for the Gauss probability math.
//    DATE: 2023-07-06


#include "Arduino.h"
#include "MultiMap.h"

#define GAUSS_LIB_VERSION       (F("0.2.0"))


class Gauss
{
public:
  Gauss()
  {
    _mean = 0;
    _reciprokeSD = 1;
  }

  //  stddev should be positive.
  bool begin(float mean = 0, float stddev = 1)
  {
    _mean = mean;
    if (stddev == 0) _reciprokeSD = NAN;
    else _reciprokeSD = 1.0 / stddev;
    return (stddev > 0);
  }


  float getMean()
  {
    return _mean;
  }


  float getStdDev()
  {
    return 1.0 / _reciprokeSD;
  }


  float P_smaller(float value)
  {
    if (_reciprokeSD == NAN) return NAN;
    return _P_smaller((value - _mean) * _reciprokeSD);
  }


  float P_larger(float value)
  {
    return 1.0 - P_smaller(value);
  }


  float P_between(float p, float q)
  {
    if (_reciprokeSD == NAN) return NAN;
    if (p >= q) return 0;
    return P_smaller(q) - P_smaller(p);
  }


  float P_outside(float p, float q)
  {
    return 1.0 - P_between(p, q);
  }


  float P_equal(float value)
  {
    if (_reciprokeSD == NAN) return NAN;
    float n = (value - _mean) * _reciprokeSD;
    //  gain of ~10% if we allocate a global var for 'constant' c
    float c = _reciprokeSD * (1.0 / sqrt(TWO_PI));
    return c * exp(-0.5 * (n * n));
  }


  float stddevs(float value)
  {
    return normalize(value);
  }


  float normalize(float value)
  {
    return (value - _mean) * _reciprokeSD;
  }


  float denormalize(float value)
  {
    return value / _reciprokeSD + _mean;
  }


  float bellCurve(float value)
  {
    return P_equal(value);
  }


  float CDF(float value)
  {
    return P_smaller(value);
  }



private:

  float _P_smaller(float x)
  {
    //  NORM.DIST(mean, stddev, x, true)
    //  these points correspond with
    //  0.0 .. 3.0 in steps of 0.1 followed by 4.0, 5.0 and 6.0 
    float __gauss[] = {
      0.50000000, 0.53982784, 0.57925971, 0.61791142,
      0.65542174, 0.69146246, 0.72574688, 0.75803635,
      0.78814460, 0.81593987, 0.84134475, 0.86433394,
      0.88493033, 0.90319952, 0.91924334, 0.93319280,
      0.94520071, 0.95543454, 0.96406968, 0.97128344,
      0.97724987, 0.98213558, 0.98609655, 0.98927589,
      0.99180246, 0.99379033, 0.99533881, 0.99653303,
      0.99744487, 0.99813419, 0.99865010, 0.99996833,
      0.99999971, 1.00000000
    };

    bool neg = false;
    if (x < 0)
    {
      neg = true;
      x = -x;
    }

    if (x >= 6.0)
    {
      if (neg) return 0.0;
      return 1.0;
    }

    if (x <= 3.0)
    {
      int idx  = x * 10;
      float rv = __gauss[idx] + ((x * 10) - idx) * (__gauss[idx+1] - __gauss[idx]);
      if (neg) return 1.0 - rv;
      return rv;
    }

    //  3.0 .. 6.0
    int xint = x;
    int idx  = 27 + xint;
    float rv = __gauss[idx] + (x - xint) * (__gauss[idx+1] - __gauss[idx]);
    if (neg) return 1.0 - rv;
    return rv;
  }

  float _mean = 0;
  //  reciprokeSD = 1.0 / stddev is faster in most math (MUL vs DIV)
  float _reciprokeSD = 1;
};


//  -- END OF FILE --

